### ISSUE TITLE
thanos query switches between "No stores registered" and normally working state back and forth

### ISSUE THREAD

**Issue by dastrobu** [thanos-io/thanos#4334]
I have setup a thanos with sidecar, store and query. 

Everything seems to be working as expected. Query shows two stores, a Sidecar and a Store. 
I can query data as expected.

After some random time stores are gone and Query shows just "No stores registered".  Sometimes stores come back after some time, but it seems to be very random. 

Logs on startup are 

```
activeStores | component | level  | msg 

- 	storeset	info	adding new storeAPI to query storeset
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	info	adding new storeAPI to query storeset
- 	storeset	warning	update of store node failed
- 	storeset	info	removing store because it's unhealthy or does not exist
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	warning	update of store node failed
- 	storeset	info	adding new storeAPI to query storeset
- 	 - 	info	changing probe status
- 	query	info	listening for requests and metrics
- 	query	info	TLS is disabled.
- 	 - 	info	changing probe status
- 	query	info	listening for serving gRPC
- 	 - 	info	starting query node
- 	 - 	info	

```

at the point where the stores are gone I don't see any specific log. When enabling debug level I just see an endless stream of 
```

activeStores | component | level  | msg 

2	storeset	debug	checked requested storeAPIs
 - 	storeset	debug	starting updating storeAPIs
2	storeset	debug	checked requested storeAPIs
 - 	storeset	debug	starting updating storeAPIs
2	storeset	debug	checked requested storeAPIs
```

I am a bit stuck here and looking for advise how to further analyze this. 
Not sure if there is kind of a corner case bug in thanos, or something wrong with my config or an issue with our infrastructure or setup. 

Config:
```
<image-entrypoint> query --http-address=0.0.0.0:9090 --log.format=json --log.level=debug --query.lookback-delta=15m --store=thanos-store.my-namespace.svc.cluster.local:10901 --store=prometheus.my-namespace.svc.cluster.local:10901
```

Version:
```
thanos --version
thanos, version 0.21.1 (branch: HEAD, revision: 3558f4a5adc8d4068f82528570a439d440f98b9d)
  build user:       circleci@eb60c35e294a
  build date:       20210604-12:24:45
  go version:       go1.16.4
  platform:         linux/amd64
```

**Comment by wiardvanrij** [comment]
The only thing I can think of is the usage of a svc as endpoint. You pass along a 'loadbalancer' which contains n-amount of services behind it. So this will swap between endpoints (I think!). 

See the usage for `--store` : 

```
Addresses of statically configured store API
                                 servers (repeatable). The scheme may be
                                 prefixed with 'dns+' or 'dnssrv+' to detect
                                 store API servers through respective DNS
                                 lookups.
```

So I would recommend changing to either direct links to each store or use dnssrv to fetch the actual endpoints. 

```
dnssrv+_{service-port-name}._{service-protocol}.{service-name}.{namespace}.cluster.local
```

I.e.:

```
--store=dnssrv+_grpc._tcp.thanos-store-grpc.monitoring.svc.cluster.local
```

(obviously change entries to your setup).

**Comment by dastrobu** [comment]
I have some updates on the issue. It seems to be a pure Thanos Query UI related issue. When I do many refreshes on the Thanos Query UI I see some logs that say `Could not get file` for some UI files. The thing is that these warn logs are not reproducible. There seems to be an issue either with serving the UI files from Thanos or with the node/filesystem/... of our cluster. 

One of my next steps will be to integrate Thanos Query into Grafana and see if I can query Thanos metrics without issues. 


Relevant log entries:
```
warn	Could not get file			pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map not found
warn	Could not get file			pkg/ui/static/react/static/css/main.28d183d2.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/main.28d183d2.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/js/main.1351c709.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/main.1351c709.chunk.js.map not found
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
warn	Could not get file			pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/css/main.28d183d2.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/main.28d183d2.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map not found
warn	Could not get file			pkg/ui/static/react/static/js/main.1351c709.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/main.1351c709.chunk.js.map not found
error	failed to parse host port for http log decision				address thanos-query-pmd-mon-dev.cnap-00-mp-dev.bmwgroup.net: missing port in address
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
error	failed to parse host port for http log decision				address thanos-query-pmd-mon-dev.cnap-00-mp-dev.bmwgroup.net: missing port in address
warn	Could not get file			pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map not found
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
warn	Could not get file			pkg/ui/static/react/static/vendor/bootstrap-4.1.3/js/bootstrap.min.js.map	AssetInfo pkg/ui/static/react/static/vendor/bootstrap-4.1.3/js/bootstrap.min.js.map not found
warn	Could not get file			pkg/ui/static/react/static/vendor/bootstrap-4.1.3/css/bootstrap.min.css.map	AssetInfo pkg/ui/static/react/static/vendor/bootstrap-4.1.3/css/bootstrap.min.css.map not found
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
debug	starting updating storeAPIs		2		
debug	checked requested storeAPIs	2	2		
warn	Could not get file			pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/2.e81d7923.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/css/main.28d183d2.chunk.css.map	AssetInfo pkg/ui/static/react/static/css/main.28d183d2.chunk.css.map not found
warn	Could not get file			pkg/ui/static/react/static/js/main.1351c709.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/main.1351c709.chunk.js.map not found
warn	Could not get file			pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map	AssetInfo pkg/ui/static/react/static/js/2.c9546e23.chunk.js.map not found
error	failed to parse host port for http log decision				address thanos-query-pmd-mon-dev.cnap-00-mp-dev.bmwgroup.net: missing port in address
```

**Comment by wiardvanrij** [comment]
Seems to be related to https://github.com/thanos-io/thanos/issues/4118 maybe

**Comment by dastrobu** [comment]
I looks like this could actually be related to #4118. 
Thanos query is accessed via an [oauth proxy](https://github.com/openshift/oauth-proxy) in our setup. Setting `-pass-host-header` to `false`, might have solved the issue. 
The fix is just hard to validate, so far I did not see the issue again. 

@wiardvanrij thanks for your input on this issue 👍🏼

**Comment by yeya24** [comment]
Should be solved in #4348. Thanks @spaparaju
